src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,906 @@
+        if (r < 0)
+                return r;
+
+        ctx->u.nft.nfnl = TAKE_PTR(nfnl);
+        return 0;
+}
+
+void fw_nftables_exit(fw_ctx *ctx) {
+        _cleanup_(sd_netlink_unrefp) sd_netlink *nfnl = NULL;
+
+        nfnl = TAKE_PTR(ctx->u.nft.nfnl);
+}
+
+static int nft_message_add_setelem_iprange(sd_netlink_message *m,
+                                           const union in_addr_union *source,
+                                           unsigned int prefixlen) {
+        uint32_t mask, start, end;
+        unsigned int nplen;
+        int r;
+
+        assert(prefixlen <= 32);
+        nplen = 32 - prefixlen;
+
+        mask = (1U << nplen) - 1U;
+        mask = htobe32(~mask);
+        start = source->in.s_addr & mask;
+
+        r = sd_nfnl_nft_message_add_setelem(m, 0, &start, sizeof(start), NULL, 0);
+        if (r < 0)
+                return r;
+
+        r = sd_nfnl_nft_message_add_setelem_end(m);
+        if (r < 0)
+                return r;
+
+        end = be32toh(start) + (1U << nplen);
+        if (end < be32toh(start))
+                end = 0U;
+        end = htobe32(end);
+
+        r = sd_nfnl_nft_message_add_setelem(m, 1, &end, sizeof(end), NULL, 0);
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_u32(m, NFTA_SET_ELEM_FLAGS, htobe32(NFT_SET_ELEM_INTERVAL_END));
+        if (r < 0)
+                return r;
+
+        r = sd_nfnl_nft_message_add_setelem_end(m);
+        if (r < 0)
+                return r;
+
+        return r;
+}
+
+#define NFT_MASQ_MSGS   3
+
+int fw_nftables_add_masquerade(
+                fw_ctx *ctx,
+                bool add,
+                int af,
+                const union in_addr_union *source,
+                unsigned int source_prefixlen) {
+        sd_netlink_message *transaction[NFT_MASQ_MSGS] = {};
+        struct sd_netlink *nfnl = ctx->u.nft.nfnl;
+        size_t tsize;
+        int r;
+
+        if (!source || source_prefixlen == 0)
+                return -EINVAL;
+
+        r = sd_nfnl_message_batch_begin(nfnl, &transaction[0]);
+        if (r < 0)
+                return r;
+        tsize = 1;
+        if (add)
+                r = sd_nfnl_nft_message_new_setelems_begin(nfnl, &transaction[tsize], af, NFT_SYSTEMD_TABLE_NAME, NFT_SYSTEMD_MASQ_SET_NAME);
+        else
+                r = sd_nfnl_nft_message_del_setelems_begin(nfnl, &transaction[tsize], af, NFT_SYSTEMD_TABLE_NAME, NFT_SYSTEMD_MASQ_SET_NAME);
+
+        if (r < 0)
+                goto out_unref;
+
+        r = nft_message_add_setelem_iprange(transaction[tsize], source, source_prefixlen);
+        if (r < 0)
+                goto out_unref;
+
+        ++tsize;
+        assert(tsize < NFT_MASQ_MSGS);
+        r = sd_nfnl_message_batch_end(nfnl, &transaction[tsize]);
+        if (r < 0)
+                return r;
+        ++tsize;
+        r = nfnl_netlink_sendv(nfnl, transaction, tsize);
+
+out_unref:
+        while (tsize)
```

Comment:
we usually avoid relying on C's downgrade-to-bool feature for numerics, but use explicit comparisons for them, to not create the impression something was used as a bool, while it is actually a number. i.e. `while (tsize > 0)…` would be better

Reply:
sorry about this, changed to while (tsize > 0) here and elsewhere.